### DISCUSSION TITLE
Accessing the intermediate noise_pred values inside a pipeline

### DISCUSSION THREAD

**Discussion by SantiagoJN** [huggingface/diffusers#7888] [Q&A]
Hi,

I'm making a small modification in the training of an IP-Adapter, and I was wondering if there was a way to obtain the value of the noise_pred that is computed in the call of a diffuser pipeline. For instance, [here](https://github.com/huggingface/diffusers/blob/b69fd990ad8026f21893499ab396d969b62bb8cc/src/diffusers/pipelines/controlnet/pipeline_controlnet_inpaint.py#L1537C17-L1537C27) is where the noise_pred is computed as output of the inner unet, and I would like to recover that value in the script where I'm calling the pipeline. Does it make sense? Is there any workaround to achieve this behavior?

Please let me know if I'm missing something! :)

**Comment by asomoza** [comment]
Hi, you can use a [callback](https://huggingface.co/docs/diffusers/v0.27.2/en/using-diffusers/callback#pipeline-callbacks) function.

**Reply by SantiagoJN** [reply]
Hi,

Thanks for the prompt response. According to the documentation, each pipeline has its own list `_callback_tensor_inputs` with the inputs that can be specified in the `callback_on_step_end_tensor_inputs` parameter of the `__call__` method. Looking at the [source code](https://github.com/huggingface/diffusers/tree/v0.27.2/src/diffusers/pipelines/controlnet) of some pipelines, I found out that they usually include the variables `["latents", "prompt_embeds", "negative_prompt_embeds"]`, but not the `noise_pred`.
Is there any way in which I can fix this?

Thank you!

**Reply by asomoza** [reply]
The `callback_on_step_end_tensor_inputs` is for when you want to return those values, in other words, when you modify them and want the pipeline to use them. 

Since you're only asking to recover the `noise_pred` I though you could do it with the pipeline object, but I just checked that the `noise_pred` is not an instance variable, I missed that.

The next step you can do is to create a custom pipeline or maybe explain a bit more why do you want to do this, maybe there's an alternative or we can add it if it's something that people could use a lot.

cc @yiyixuxu

**Reply by SantiagoJN** [reply]
I'm trying to train a modified version of an [IP-Adapter](https://ip-adapter.github.io/). In their [official repository](https://github.com/tencent-ailab/IP-Adapter/tree/main), they provide some [sample code](https://github.com/tencent-ailab/IP-Adapter/blob/main/tutorial_train.py) on how to train their models, where basically they have some trainable parameters whose weights are learnt through a [MSE loss computed between noises](https://github.com/tencent-ailab/IP-Adapter/blob/11a904682126e5dfdd0c1ce86fdb435e9e01c1e0/tutorial_train.py#L389). The _problem_ is that in this implementation, they simply use the [unet](https://github.com/tencent-ailab/IP-Adapter/blob/11a904682126e5dfdd0c1ce86fdb435e9e01c1e0/tutorial_train.py#L120) to make the forward pass, but I want to use instead some diffuser's pipeline, like [controlnet_inpaint one](https://github.com/huggingface/diffusers/blob/v0.27.2/src/diffusers/pipelines/controlnet/pipeline_controlnet_inpaint.py), to compute the `noise_pred` at some timestep.

I'm not really sure if this is something that people could use a lot. Anyways, creating a custom pipeline sounds like a good approach, I'll give it a try!

Thank you so much for the help!
